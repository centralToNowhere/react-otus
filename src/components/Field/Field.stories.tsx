import React from "react";
import { Story, Meta } from "@storybook/react";
import { Field, IFieldProps } from "./Field";
import figures from "./figures.json";

export default {
  title: "Components/Field",
  component: Field,
  args: {
    cellSize: 40,
    cellsBetweenChars: 1,
    inputString: "123",
    figuresMap: figures["figures"],
    fieldWidthPx: 600,
    fieldHeightPX: 400,
  },
} as Meta;

const Template: Story<IFieldProps> = (args) => <Field {...args} />;

export const FieldSmall = Template.bind({});

FieldSmall.args = {
  cellSize: 20,
};

export const FieldBig = Template.bind({});

FieldBig.args = {
  cellSize: 10,
};
