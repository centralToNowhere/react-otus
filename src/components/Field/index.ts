export { Field } from "./Field";
export type { IFiguresMap, IFieldProps } from "./Field";
