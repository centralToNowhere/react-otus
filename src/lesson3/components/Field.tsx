import React from "react";
import "./field.css";
import ListCells from "./ListCells";
import { ICell, isCell } from "./Cell";

interface FieldProps {
  rowSize?: number;
  backgroundColor?: string;
  inputCells?: ICell[][];
}

interface FieldSizes {
  cellSize: number;
  fieldWidth: number;
}

export const getRandomColor = (): string => {
  const getRandNumber256 = (): number => {
    return parseInt(String(Math.random() * 256));
  };

  return ["r", "g", "b"].reduce((a, b, i) => {
    return `${a}${i < 2 ? `${getRandNumber256()},` : `${getRandNumber256()})`}`;
  }, "rgb(");
};

export const getRandomCellsByRowSize = (rowSize: number): ICell[][] => {
  const arr: ICell[][] = [];

  for (let i = 0; i < rowSize; i++) {
    for (let j = 0; j < rowSize; j++) {
      const cell: ICell = {
        number: i * rowSize + j + 1,
        backgroundColor: getRandomColor(),
        alive: Boolean(Math.round(Math.random())),
      };

      if (!arr[i]) {
        arr[i] = [];
      }

      arr[i][j] = cell;
    }
  }

  return arr;
};

export const isValidCells = (cells: [][] | ICell[][]): boolean => {
  if (!Array.isArray(cells)) {
    return false;
  }

  cells = cells as []; // cells probably an array here
  const sideLength = cells.length;

  if (sideLength === 0) {
    return false;
  }

  for (let i = 0; i < sideLength; i++) {
    if (!Array.isArray(cells[i]) || cells[i].length !== sideLength) {
      return false;
    }

    cells[i] = cells[i] as [];

    for (let j = 0; j < sideLength; j++) {
      if (!isCell(cells[i][j])) {
        return false;
      }
    }
  }

  return true;
};

const calcFieldWidth = (
  cellSize: number,
  cellsNumberInRow: number,
  padding: number
): number => {
  return (cellSize + padding * 2) * cellsNumberInRow + padding * 2;
};

const calcCellSize = (
  padding: number,
  maxCellSize: number,
  cellsNumberInRow: number
): FieldSizes => {
  const cellPadding = padding;
  const outerPadding = padding;
  const outerMargin = 20;
  const screenWidth = window.innerWidth;
  const screenHeight = window.innerHeight;
  const minSide = Math.min(screenWidth, screenHeight);
  const fieldWidth = calcFieldWidth(maxCellSize, cellsNumberInRow, padding);

  // overflow, need recalculate
  if (minSide < fieldWidth + 2 * outerMargin) {
    const cellSize =
      (minSide -
        2 * outerMargin -
        2 * outerPadding -
        (cellPadding * 2 + cellPadding * cellsNumberInRow * 2)) /
      cellsNumberInRow;

    return {
      cellSize: cellSize,
      fieldWidth: calcFieldWidth(cellSize, cellsNumberInRow, padding),
    };
  }

  return {
    // no overflow, good
    cellSize: maxCellSize,
    fieldWidth: fieldWidth,
  };
};

const Field = (props: FieldProps) => {
  const { rowSize = 10, backgroundColor, inputCells = [[]] } = props;

  const cells = isValidCells(inputCells)
    ? inputCells
    : getRandomCellsByRowSize(rowSize);
  const cellPadding = 5;
  const maxCellSize = 50;
  const { cellSize, fieldWidth } = calcCellSize(
    cellPadding,
    maxCellSize,
    rowSize
  );

  const fieldStyles = {
    backgroundColor,
    width: fieldWidth,
  };

  return (
    <div className="container">
      <div className="field" data-testid="field" style={fieldStyles}>
        <ListCells
          cells={cells}
          cellSize={cellSize}
          cellPadding={cellPadding}
        />
      </div>
    </div>
  );
};

export default Field;
