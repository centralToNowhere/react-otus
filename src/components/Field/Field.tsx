import React from "react";
import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";
import { Cell, ICell, isCell } from "@/components/Cell";

export interface IFieldProps {
  cellSize: number;
  activeCells?: ICell[];
  cellsInRow: number;
  cellsInCol: number;
}

type FieldProps = typeof Field.defaultProps & IFieldProps;

export class Field extends React.Component<FieldProps> {
  static defaultProps = {
    activeCells: [] as ICell[],
  };

  constructor(props: FieldProps) {
    super(props);

    this.isActiveCell = this.isActiveCell.bind(this);
  }

  isActiveCell(cell: ICell): boolean {
    return !!this.props.activeCells.find(
      (activeCell: ICell) =>
        isCell(activeCell) && activeCell.x === cell.x && activeCell.y === cell.y
    );
  }

  renderCells() {
    const cellsInCol = this.props.cellsInCol;
    const cellsInRow = this.props.cellsInRow;

    return Array.from(Array(cellsInCol)).map((col, i) =>
      Array.from(Array(cellsInRow)).map((row, j) => (
        <Cell
          key={(i + 1) * (j + 1)}
          size={this.props.cellSize}
          isActive={this.isActiveCell({ x: j, y: i })}
          number={(i + 1) * cellsInRow - (cellsInRow - (j + 1))}
        />
      ))
    );
  }

  shouldComponentUpdate(nextProps: Readonly<IFieldProps>): boolean {
    if (nextProps.cellSize < 10) {
      return false;
    }

    return true;
  }

  render() {
    return (
      <FieldContainer
        cellSize={this.props.cellSize}
        cellsInRow={this.props.cellsInRow}
        cellsInCol={this.props.cellsInCol}
        data-testid={"field"}
      >
        {this.renderCells()}
      </FieldContainer>
    );
  }
}

interface IFieldStyledContainerProps {
  cellSize: number;
  cellsInRow: number;
  cellsInCol: number;
}

const FieldContainer = styled.div<IFieldStyledContainerProps>`
  width: ${(props) => props.cellsInRow * props.cellSize}px;
  border-bottom: 1px solid black;
  outline: 1px solid black;
  cursor: pointer;
  height: ${(props) => props.cellsInCol * props.cellSize}px;
  display: flex;
  flex-wrap: wrap;
  background: ${COLORS.primary};
  margin: 0 auto;
`;
