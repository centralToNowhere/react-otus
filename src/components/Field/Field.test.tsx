import React from "react";
import { render, screen } from "@testing-library/react";
import { Field, IFiguresMap } from "./Field";
import figures from "./figures.json";

const cellSize = 40;
const cellsBetweenChars = 1;
const fieldWidth = 600;
const fieldHeight = 400;

interface ICharsIndexes {
  [key: string]: number[];
}

const charsIndexes: ICharsIndexes = {
  "1": [
    38, 39, 52, 53, 54, 68, 69, 83, 84, 98, 99, 113, 114, 127, 128, 129, 130,
  ],
  "2": [
    37, 38, 39, 40, 51, 52, 55, 56, 70, 71, 83, 84, 85, 98, 97, 111, 112, 126,
    127, 128, 129, 130, 131,
  ],
  "3": [
    37, 38, 39, 40, 51, 52, 55, 56, 70, 71, 83, 84, 85, 100, 101, 111, 112, 115,
    116, 127, 128, 129, 130,
  ],
  "4": [
    36, 37, 40, 41, 51, 52, 55, 56, 66, 67, 70, 71, 81, 82, 85, 86, 97, 98, 99,
    100, 101, 115, 116, 130, 131,
  ],
  "5": [
    36, 37, 38, 39, 40, 41, 51, 52, 66, 67, 68, 69, 70, 85, 86, 100, 101, 115,
    116, 111, 112, 127, 128, 129, 130,
  ],
  "6": [
    37, 38, 39, 40, 51, 52, 55, 56, 66, 67, 81, 82, 83, 84, 85, 96, 97, 100,
    101, 111, 112, 115, 116, 127, 128, 129, 130,
  ],
  "7": [
    36, 37, 38, 39, 40, 41, 55, 56, 69, 70, 83, 84, 98, 99, 113, 114, 128, 129,
  ],
  "8": [
    37, 38, 39, 40, 51, 52, 55, 56, 66, 67, 70, 71, 82, 83, 84, 85, 96, 97, 100,
    101, 111, 112, 115, 116, 127, 128, 129, 130,
  ],
  "9": [
    37, 38, 39, 40, 51, 52, 55, 56, 66, 67, 70, 71, 82, 83, 84, 85, 100, 101,
    111, 112, 115, 116, 127, 128, 129, 130,
  ],
  "0": [
    37, 38, 39, 40, 51, 52, 55, 56, 66, 67, 70, 71, 81, 82, 85, 86, 96, 97, 100,
    101, 111, 112, 115, 116, 127, 128, 129, 130,
  ],
};

const testCellsToBeActive = (inputString: string): void => {
  render(
    <Field
      cellSize={cellSize}
      cellsBetweenChars={cellsBetweenChars}
      inputString={inputString}
      figuresMap={figures["figures"] as IFiguresMap["figures"]}
      fieldWidthPx={fieldWidth}
      fieldHeightPX={fieldHeight}
    />
  );

  const cells: HTMLElement[] = screen.getAllByTestId("cell");
  const activeCellsIndexes = charsIndexes[inputString];

  activeCellsIndexes.forEach((cellIndex) => {
    expect(cells[cellIndex - 1]).toHaveStyle({
      background: "black",
    });
  });
};

describe("Field tests", () => {
  it("should render field", () => {
    const { asFragment } = render(
      <Field
        cellSize={30}
        cellsBetweenChars={1}
        inputString={""}
        figuresMap={figures["figures"] as IFiguresMap["figures"]}
        fieldWidthPx={600}
        fieldHeightPX={400}
      />
    );

    const field = screen.getByTestId("field");

    expect(field).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 150 inactive cells", () => {
    const inputString = "";

    render(
      <Field
        cellSize={cellSize}
        cellsBetweenChars={cellsBetweenChars}
        inputString={inputString}
        figuresMap={figures["figures"] as IFiguresMap["figures"]}
        fieldWidthPx={fieldWidth}
        fieldHeightPX={fieldHeight}
      />
    );

    const cells: HTMLElement[] = screen.getAllByTestId("cell");
    expect(cells.length).toBe(150);

    cells.forEach((cell) => {
      expect(cell).not.toHaveStyle({
        background: "black",
      });
    });
  });

  it("should render 1", () => {
    testCellsToBeActive("1");
  });

  it("should render 2", () => {
    testCellsToBeActive("2");
  });

  it("should render 3", () => {
    testCellsToBeActive("3");
  });

  it("should render 4", () => {
    testCellsToBeActive("4");
  });

  it("should render 5", () => {
    testCellsToBeActive("5");
  });

  it("should render 6", () => {
    testCellsToBeActive("6");
  });

  it("should render 7", () => {
    testCellsToBeActive("7");
  });

  it("should render 8", () => {
    testCellsToBeActive("8");
  });

  it("should render 9", () => {
    testCellsToBeActive("9");
  });

  it("should render 0", () => {
    testCellsToBeActive("0");
  });
});
