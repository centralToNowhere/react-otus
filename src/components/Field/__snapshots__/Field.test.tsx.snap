// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field tests should render field 1`] = `
<DocumentFragment>
  <div
    class="css-jy02vj"
    data-testid="field"
  >
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
    <div
      class="css-11pa12e"
      data-testid="cell"
      size="30"
    />
  </div>
</DocumentFragment>
`;
